File: Library/knoics/RubiksCube/BasicOp.cs
#region Description
//-----------------------------------------------------------------------------
// File:        BasicOp.cs
// Author:      Ligang Wang
// Email:       ligang@dingn.com
// Date:        01/31/2009
//-----------------------------------------------------------------------------
#endregion
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Knoics.Math;

namespace Knoics.RubiksCube
{

    public class CubicleOrientation
    {
        public string Name { get; set; }
        public string OrientationName { get; set; }

        public CubicleOrientation(string name, string orientationName)
        {
            Name = name;
            OrientationName = orientationName;
        }
    }

    public class OpFromBasic
    {
        public OpFromBasic(string op, string basicOp, RotationDirection dir)
        {
            Op = op;
            BasicOp = basicOp;
            RotationDirectionWithAxis = dir;
        }
        public string Op { get; set; }
        public string BasicOp { get; set; }
        public RotationDirection RotationDirectionWithAxis { get; set; }
    }
    

    public class BasicOp
    {
        public string Op { get; set; }
        public Axis Axis { get; set; }
        public double RotationAngle { get; set; }

        public BasicOp(string op, string[][] cycles, Axis axis, double angle)
        {
            Op = op;
            CubicleOrientationCycles = cycles;
            Axis = axis;
            RotationAngle = angle;
        }

        public string[][] CubicleOrientationCycles
        {
            set
            {
                _cubicleGroupCycles = new CubicleOrientation[value.Length][];
                _cubicleGroup = new List<string>();
                int i = 0;
                foreach (string[] orientationNames in value)
                {
                    _cubicleGroupCycles[i] = new CubicleOrientation[orientationNames.Count()];
                    int j = 0;
                    foreach (string orientationName in orientationNames)
                    {
                        CubicleOrientation co = new CubicleOrientation(GetNameFromOrientedName(orientationName), orientationName);
                        _cubicleGroupCycles[i][j++] = co;
                        _cubicleGroup.Add(co.Name);
                    }
                    i++;
                }
            }
        }


        private static string GetNameFromOrientedName(string orientationName)
        {
            string name = orientationName;
            int pos = orientationName.IndexOfAny(new char[] { 'U', 'D' });
            if (pos >= 0)
            {
                name = orientationName.Substring(pos, orientationName.Length - pos) + orientationName.Substring(0, pos);
            }
            else 
            {
                pos = orientationName.IndexOfAny(new char[] { 'L', 'R' });
                if (pos > 0)
                {
                    name = orientationName.Substring(pos, orientationName.Length - pos) + orientationName.Substring(0, pos);
                }
            }
            return name;
        }

        //types of cycle, cycle elements
        private CubicleOrientation[][] _cubicleGroupCycles;
        public CubicleOrientation[][] CubicleGroupCycles
        {
            get
            {
                return _cubicleGroupCycles;
            }
        }


        private List<string> _cubicleGroup;
        public List<string> CubicleGroup
        {
            get { return _cubicleGroup; }
        }
    }

}

File: Knoics/BasicOp.fs
//-----------------------------------------------------------------------------
// File:        BasicOp.fs
// Author:      Ligang Wang
// Email:       ligang@dingn.com
// Date:        07/25/2009
//-----------------------------------------------------------------------------
namespace Knoics.RubiksCube

open System
open Knoics.Math
open System.Collections.Generic

type CubicleOrientation(name:string, orientationName:string) =
    member x.Name = name
    member x.OrientationName = orientationName

type OpFromBasic(op:string, basicOp:string, rotationDirectionWithAxis:RotationDirection) = 
    member x.Op = op
    member x.BasicOp = basicOp
    member x.RotationDirectionWithAxis  = rotationDirectionWithAxis
    
//cubicleOrientationCycles, cycles for corner cubicles, edge cubicles and center cubicles
type BasicOp(op:string, cubicleOrientationCycles:string array array, axis:Axis, angle:double) =
    let GetNameFromOrientedName(orientationName:string) = 
        let name = orientationName;
        let pos = orientationName.IndexOfAny([|'U';'D'|])//(new char[] { 'U', 'D' })
        if (pos >= 0) then
            orientationName.Substring(pos, orientationName.Length - pos) + orientationName.Substring(0, pos)
        else 
            let pos1 = orientationName.IndexOfAny([|'L'; 'R'|])
            if (pos1 > 0) then
                orientationName.Substring(pos1, orientationName.Length - pos1) + orientationName.Substring(0, pos1)
            else
                orientationName

    let _cubicleGroupCycles = cubicleOrientationCycles |> Array.map (fun arr -> (arr |> Array.map( fun e -> new CubicleOrientation(GetNameFromOrientedName(e), e)))) //new CubicleOrientation[cubicleOrientationCycles.Length][]
    let _cubicleGroup = [for arr in cubicleOrientationCycles do
                            for s in arr -> GetNameFromOrientedName(s)
                         ] 
    member x.Op = op
    member x.Axis = axis
    member x.RotationAngle = angle
    member x.CubicleGroupCycles with get() = _cubicleGroupCycles
    member x.CubicleGroup with get() = _cubicleGroup

